### PR TITLE
[feature](file reader) Merge hdfs reader to the new file reader

### LINKED ISSUE #15456: [Enhancement](file reader) Refactor file reader
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.

### Description

Currently, there are two sets of file readers in Doris. We need to integrate the old file reader into the new file reader.

The purposes of the new file reader are:
1. Simplifies the file reader interface, and abstracts the file system to facilitate access to different storage media.
    Accessing different storage media will have different file systems (except `streamLoadPipe`), and file readers are 
    generated from corresponding file system。
2. Prepare for the introduction of File Cache in the future.
     We will import file cache for accessing `RemoteFileSystem` (such as HdfsFileSystem and S3FileSystem), and the logic of the file cache will be hidden inside the file reader so that for the upper-layer caller, the change of the file cache does not need to modify the upper-layer calling logic

### Solution

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by BePPPower**
# Proposed changes

Issue Number: close #15456 

Currently, there are two sets of file readers in Doris, this pr rewrites the old hdfs file reader with the new file reader.

TODO:
1. rewrite broker reader
2. rewrite stream load pipe and kafka consumer pipe

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
4. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [x] No Need
5. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [x] No Need
6. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
7. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.38 seconds
 load time: 426 seconds
 storage size: 17123356275 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221207072948_clickbench_pr_59162.html

File: be/src/io/fs/hdfs_file_system.h [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+    uint64 _hdfs_hash_code(THdfsParams& hdfs_params);
+    Status _create_fs(THdfsParams& hdfs_params, hdfsFS* fs);
+    void _clean_invalid();
+    void _clean_oldest();
+};
+
+class HdfsFileSystem final : public RemoteFileSystem {
+public:
+    HdfsFileSystem(THdfsParams hdfs_params, const std::string& path);
```
Comment (platoneko) [inline be/src/io/fs/hdfs_file_system.h:112]: using `const THdfsParams&` is better.

File: be/src/io/fs/hdfs_file_system.h [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    uint64_t _now() {
+        return std::chrono::duration_cast<std::chrono::milliseconds>(
+                       std::chrono::system_clock::now().time_since_epoch())
+                .count();
+    }
+};
+
+// Cache for HdfsFileSystemHandle
+class HdfsFileSystemCache {
+public:
```
Comment (platoneko) [inline be/src/io/fs/hdfs_file_system.h:84]: If `HdfsFileSystemCache` won't be used outside hdfs file system, had better define this class in .cpp

File: be/src/io/fs/hdfs_file_system.cpp [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+}
+
+HdfsFileSystem::~HdfsFileSystem() {
+    if (_fs_handle && _fs_handle->from_cache) {
+        _fs_handle->dec_ref();
+    }
+}
+
+Status HdfsFileSystem::connect() {
+    std::lock_guard lock(_handle_mu);
```
Comment (platoneko) [inline be/src/io/fs/hdfs_file_system.cpp:59]: If we never reconnect HdfsFileSystem, no need to use mutex.

File: be/src/io/fs/hdfs_file_system.h (hunk 1 above)
Comment (BePPPower) [reply be/src/io/fs/hdfs_file_system.h]: done

File: be/src/io/hdfs_reader_writer.h [hunk 4]
```diff
@@ -39,6 +44,8 @@ class HdfsReaderWriter {
 
     static Status create_writer(const std::map<std::string, std::string>& properties,
                                 const std::string& path, std::unique_ptr<FileWriter>& writer);
-};
 
+    static Status create_new_reader(const THdfsParams& hdfs_params, const std::string& path,
```
[the remark this answers is not in the document]
Comment (BePPPower) [reply be/src/io/hdfs_reader_writer.h]: done

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**